test/apps/iodemo/io_demo.cc

```diff
@@ -136,11 +171,41 @@ class MemoryPool {
     }
 
 private:
-    std::vector<T*> _free_stack;
-    std::queue<T*>  _offcache_queue;
-    uint32_t        _num_allocated;
-    size_t          _buffer_size;
-    std::string     _name;
+    size_t                   _buffer_size;
+    std::vector<BufferType*> _free_stack;
+    std::queue<BufferType*>  _offcache_queue;
+    uint32_t                 _num_allocated;
+    std::string              _name;
+};
+
+template<class BufferType, bool use_offcache = false>
+class MemoryPool : public BaseMemoryPool<BufferType, use_offcache> {
+    using BaseMemoryPool<BufferType, use_offcache>::BaseMemoryPool;
+
+public:
+    virtual BufferType *construct() override
+    {
+        return new BufferType(this->get_buffer_size(), *this);
+    }
+};
+
+template<typename BufferType>
+class BufferMemoryPool : public BaseMemoryPool<BufferType, true> {
+public:
+    BufferMemoryPool(size_t buffer_size, const std::string &name,
+                     ucs_memory_type_t memory_type, size_t offcache = 0) :
+        BaseMemoryPool<BufferType, true>(buffer_size, name, offcache),
+        _memory_type(memory_type)
+    {
+    }
+
+    virtual BufferType *construct() override
+    {
+        return new BufferType(this->get_buffer_size(), *this, _memory_type);
```

Comment:
minor: just `get_buffer_size()`

Reply:
Nope, that's impossible because of templatization & inheritance. Removing `this->` would lead to:

```
io_demo.cc: In member function ‘virtual BufferType* MemoryPool<BufferType, use_offcache>::construct()’:
io_demo.cc:193:31: error: there are no arguments to ‘get_buffer_size’ that depend on a template parameter, so a declaration of ‘get_buffer_size’ must be available [-fpermissive]
  193 |         return new BufferType(get_buffer_size(), *this);
...
```

Reply:
ok